File: Python/ConvertPathToTrajectory/GenerateDynamicObstacles_unstructured.py
import numpy as np

class Obstacle:
    def __init__(self, x, y):
        self.x = np.asarray(x)
        self.y = np.asarray(y)

    def point_at(self, i):
        return np.asarray([self.x[i], self.y[i]])   


def GenerateDynamicObstacles_unstructured(config):
    st_graph_search_ = config.st_graph_search_
    Nobs_dyn = 5

    start_obs = []
    end_obs = []

    start_obs.append(Obstacle([10,15,15,10], [10,10,15,15]))
    end_obs.append(Obstacle(  [45,50,50,45], [10,10,15,15]))

    start_obs.append(Obstacle([45,50,50,45], [10,10,15,15]))
    end_obs.append(Obstacle(  [45,50,50,45], [45,45,50,50]))

    start_obs.append(Obstacle([45,50,50,45], [45,45,50,50]))
    end_obs.append(Obstacle(  [10,15,15,10], [45,45,50,50]))

    start_obs.append(Obstacle([10,15,15,10], [45,45,50,50]))
    end_obs.append(Obstacle(  [10,15,15,10], [10,10,15,15]))

    start_obs.append(Obstacle([17,19,25,23], [19,17,23,25]))
    end_obs.append(Obstacle(  [34,36,42,40], [36,34,40,42]))

    obstacle_map = []
    for i in range(Nobs_dyn):
        dx = end_obs[i].x - start_obs[i].x
        dy = end_obs[i].y - start_obs[i].y

        obstacle_map.append([])
        for j in range(st_graph_search_.num_nodes_t):
            temp_obs = Obstacle(
                start_obs[i].x + dx / st_graph_search_.num_nodes_t * j, 
                start_obs[i].y + dy / st_graph_search_.num_nodes_t * j
                )
            obstacle_map[i].append(temp_obs)
    
    return obstacle_map


if __name__ == "__main__":
    from types import SimpleNamespace
    st_graph_search_ = SimpleNamespace(num_nodes_t=10)

    obs_map = GenerateDynamicObstacles_unstructured()

    print(obs_map)
    print(np.asarray(obs_map).shape)

File: Matlab/ConvertPathToTrajectory/GenerateDynamicObstacles_unstructured.m
function obstacle_cell = GenerateDynamicObstacles_unstructured()
global st_graph_search_  
Nobs_dyn = 5;
dynamic_obs = cell(2,5); % 结构体，第一行存储动态障碍物的起点位置，第二行存储动态障碍物的终点位置
% 第1个障碍物的起点
dynamic_obs{1,1}.x=[10,15,15,10]; dynamic_obs{1,1}.y=[10,10,15,15];
dynamic_obs{2,1}.x=[45,50,50,45]; dynamic_obs{2,1}.y=[10,10,15,15];
% 第2个障碍物的起点
dynamic_obs{1,2}.x=[45,50,50,45]; dynamic_obs{1,2}.y=[10,10,15,15];
dynamic_obs{2,2}.x=[45,50,50,45]; dynamic_obs{2,2}.y=[45,45,50,50];
% 第3个障碍物的起点
dynamic_obs{1,3}.x=[45,50,50,45]; dynamic_obs{1,3}.y=[45,45,50,50];
dynamic_obs{2,3}.x=[10,15,15,10]; dynamic_obs{2,3}.y=[45,45,50,50];
% 第4个障碍物的起点
dynamic_obs{1,4}.x=[10,15,15,10]; dynamic_obs{1,4}.y=[45,45,50,50];
dynamic_obs{2,4}.x=[10,15,15,10]; dynamic_obs{2,4}.y=[10,10,15,15];
% 第5个障碍物的起点
dynamic_obs{1,5}.x=[17,19,25,23]; dynamic_obs{1,5}.y=[19,17,23,25];
dynamic_obs{2,5}.x=[34,36,42,40]; dynamic_obs{2,5}.y=[36,34,40,42];

obstacle_cell = cell(st_graph_search_.num_nodes_t, Nobs_dyn);
for ii = 1 : Nobs_dyn
    dx = dynamic_obs{end,ii}.x(1) - dynamic_obs{1,ii}.x(1);
    dy = dynamic_obs{end,ii}.y(1) - dynamic_obs{1,ii}.y(1);
    for jj = 1 : st_graph_search_.num_nodes_t
        temp.x = dynamic_obs{1,ii}.x + dx / st_graph_search_.num_nodes_t * (jj - 1);
        temp.y = dynamic_obs{1,ii}.y + dy / st_graph_search_.num_nodes_t * (jj - 1);
        obstacle_cell{jj, ii} = temp;
    end
end
end